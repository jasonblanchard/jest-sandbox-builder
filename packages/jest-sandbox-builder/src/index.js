import get from 'lodash.get';
import set from 'lodash.set';

// TODO: Split up this file.

const registry = {};
const stack = [];
let lastIt;

export function describe(name, fn) {
  if (stack[stack.length - 1]) {
    registry[stack[stack.length - 1]][name] = {};
  } else {
    registry[name] = {};
  }
  stack.push(name);
  fn();
  stack.pop();
}

export function it(name, fn) {
  set(registry, [...stack, name], null);
  lastIt = name;
  fn();
}

export function beforeEach(fn) {
  set(registry, [...stack, 'beforeEach'], null);
  lastIt = 'beforeEach';
  fn();
}

export function factory(factoryFn, renderFn) {
  const component = factoryFn();
  set(registry, [...stack, lastIt], {
    isLeaf: true,
    component
  });

  if (process.env.NODE_ENV === 'test') return renderFn(component);

  // TODO: Abstract and generalize
  return {
    findWhere: () => ({
      exists: () => {},
      simulate: () => {}
    })
  };
}

export function getRegistry() {
  return registry;
}

// TODO: Mock all Jest globals https://facebook.github.io/jest/docs/en/api.html
export function expect() {
  return {
    toMatchSnapshot: () => {},
    toEqual: () => {},
    toBeCalled: () => {}
  };
}

export const jest = {
  fn: () => {}
}
