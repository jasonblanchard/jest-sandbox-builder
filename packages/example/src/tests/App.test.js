import React from 'react';
import App from '../App';
import { mount } from 'enzyme';

import { factory } from 'jest-sandbox-builder';

function render(props) {
  return factory(() => {
    return <App {...props} />;
  }, mount);
}

describe('App', () => {
  it('renders', () => {
    const wrapper = render();
    expect(wrapper).toMatchSnapshot();
  });
  
  describe('block with beforeEach', () => {
    let wrapper;
    
    beforeEach(() => {
      wrapper = render({ name: 'before each' });  
    });
    
    it('does something', () => {
      expect(wrapper.findWhere(element => element.type() && element.text() === 'Hello before each').exists()).toEqual(true);
    });
    
    it('does something again', () => {
      expect(wrapper.findWhere(element => element.type() && element.text() === 'Hello before each').exists()).toEqual(true);
    });
  })

  describe('prop name', () => {
    it('has a default prop name', () => {
      const wrapper = render();
      expect(wrapper.findWhere(element => element.type() && element.text() === 'Hello world').exists()).toEqual(true);
    });

    it('takes a custom prop name', () => {
      const wrapper = render({ name: 'Test World' });
      expect(wrapper.findWhere(element => element.type() && element.text() === 'Hello Test World').exists()).toEqual(true);
    });
  });

  it('calls onClick', () => {
    const onClick = jest.fn();
    const wrapper = render({ onClick });
    wrapper.findWhere(element => element.type() && element.text() === 'click me').simulate('click');

    expect(onClick).toBeCalled();
  });

  it('Does not include a render', () => {
    expect(1).toEqual(1);
  })
});

it('Outside App test main describe', () => {
  const wrapper = render();
  expect(1).toEqual(1);
});
