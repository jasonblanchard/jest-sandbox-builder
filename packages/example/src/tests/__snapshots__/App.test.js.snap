// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `
<App
  name="world"
  onClick={[Function]}
>
  <div
    className="App"
  >
    <h2
      className="App-title"
    >
      Hello 
      world
    </h2>
    <Nested
      onClick={[Function]}
    >
      <div
        className="Nested"
      >
        This is a nested component 
        <button
          onClick={[Function]}
        >
          click me
        </button>
      </div>
    </Nested>
  </div>
</App>
`;
